src/appleseed.studio/mainwindow/pythonconsole/pythonconsolewidget.cpp

```diff
@@ -121,14 +165,166 @@ void PythonConsoleWidget::slot_execute_all()
+        msgbox.setDefaultButton(QMessageBox::Save);
+        return msgbox.exec();
+    }
+}
+
+bool PythonConsoleWidget::can_close_file()
+{
+    // Unmodified file: no problem.
+    if (!m_is_file_dirty)
+        return true;
+
+    // The current file has been modified, ask the user what to do.
+    switch (show_modified_file_message_box(this))
+    {
+        case QMessageBox::Save:
+            slot_save_file();
+            return true;
+
+        case QMessageBox::Discard:
+            return true;
+
+        case QMessageBox::Cancel:
+            return false;
+    }
+
+    assert(!"Should never be reached.");
+    return false;
+}
+
+void PythonConsoleWidget::slot_open_file()
+{
+    if (!can_close_file())
+        return;
+
+    ParamArray settings = PythonInterpreter::instance().get_main_window()->get_settings();
+
+    QString filepath =
+        get_open_filename(
+            this,
+            "Open...",
+            "Python Script File (*.py)",
+            settings,
+            SETTINGS_FILE_DIALOG_PYTHON_SCRIPTS);
+
+    if (!filepath.isEmpty())
+    {
+        filepath = QDir::toNativeSeparators(filepath);
+        open_file(filepath.toStdString());
+    }
+}
+
+void PythonConsoleWidget::open_file(const string& filepath)
+{
+    fstream file(filepath, fstream::in);
+    if (file.bad())
+    {
+        RENDERER_LOG_ERROR("Can't open python script \"%s\"", filepath.c_str());
+        return;
+    }
+
+    close_file();
+    m_opened_file = filepath;
+
+    while (!file.eof())
+    {
+        string str;
+        getline(file, str);
+        m_input->appendPlainText(str.c_str());
+    }
+
+    file.close();
+    m_is_file_dirty = false;
+}
+
+void PythonConsoleWidget::slot_save_file()
+{
+    if (!has_file_path())
+        slot_save_file_as();
+    else
+        save_file(m_opened_file);
+}
+
+bool PythonConsoleWidget::has_file_path()
+{
+    return m_opened_file != "";
+}
+
+void PythonConsoleWidget::slot_save_file_as()
+{
+    ParamArray settings = PythonInterpreter::instance().get_main_window()->get_settings();
+
+    const QString filepath =
+        get_save_filename(
+            this,
+            "Save As...",
+            "Python Script File (*.py)",
+            settings,
+            SETTINGS_FILE_DIALOG_PYTHON_SCRIPTS);
+
+    if (!filepath.isEmpty())
+        save_file(filepath.toStdString());
+}
+
+void PythonConsoleWidget::save_file(std::string filepath)
+{
+    const size_t extension_start = filepath.rfind('.') + 1;
+    string extension = "";
+    if (extension_start != string::npos)
+        extension = filepath.substr(extension_start);
+
+    if (extension != "py")
+        filepath += ".py";
+
+    fstream file(filepath, fstream::out);
+    if (file.bad())
+    {
+        RENDERER_LOG_ERROR("Can't save python script \"%s\"", filepath.c_str());
+        return;
+    }
+
+    string text = m_input->toPlainText().toStdString();
+    file.write(text.c_str(), text.size());
```

Comment:
maybe. This << operator is sometimes smarter than needed. Will it work correctly here?